### PR TITLE
feat: User Permissions

### PR DESCRIPTION
**Pull request by cbolles**
# Description

* Add basic user permission checks
* Update middleware to new Swift S3 version
* Mongoose added to server

## Checklist

- [x] This PR can be reviewed in under 30 minutes
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have assigned reviewers to this PR.

### CODE REVIEW

File: middleware/sailauth.py [hunk 1]
```diff
@@ -132,7 +134,7 @@ def handle_auth(self, env, start_response):
             'x-auth-token': provided_key,
             'x-storage-token': provided_key,
             'x-auth-token-expires': '86400',
-            'x-storage-url': 'http://127.0.0.1:12345/v1/{}'.format(account),
+            'x-storage-url': 'http://127.0.0.1:8080/v1/{}'.format(account),
```
Comment (gregfrasco) [inline middleware/sailauth.py:137]: I this hard coded on purpose or should this be an env variable?

File: middleware/sailauth.py [hunk 2]
```diff
@@ -179,7 +223,12 @@ def authorize(self, req):
         if token is None:
             return HTTPUnauthorized(request=req, body='No token provided')
 
-        req.environ['PATH_INFO'] = self.get_updated_path(req.environ['PATH_INFO'], self.get_account(req.environ))
+        # TODO: When evaulating how to handle accounts, update this
+        # req.environ['PATH_INFO'] = self.get_updated_path(req.environ['PATH_INFO'], self.get_account(req.environ))
+        print('ENV', req.environ)
```
Comment (gregfrasco) [inline middleware/sailauth.py:228]: Do we need this print

File: server/src/app.module.ts [hunk 3]
```diff
@@ -10,7 +12,9 @@ import { ApolloDriver } from '@nestjs/apollo';
     GraphQLModule.forRoot({
       autoSchemaFile: 'schema.gql',
       driver: ApolloDriver
-    })
+    }),
+    MongooseModule.forRoot('mongodb://localhost:27017/nist'),
```
Comment (gregfrasco) [inline server/src/app.module.ts:16]: Could we make this pull from config service?

File: server/src/auth/dtos/resource.dto.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { InputType, Field, registerEnumType } from '@nestjs/graphql';
+import { RequestMethod } from '@nestjs/common';
+
+registerEnumType(RequestMethod, { name: 'RequestMethod' });
```
Comment (gregfrasco) [inline server/src/auth/dtos/resource.dto.ts:4]: Nice

File: server/src/auth/user-permissions.service.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      return false;
+    }
+
+    // Cannot delete a bucket
+    // TODO: Limit this to the root level account
+    if (request.method == RequestMethod.DELETE) {
+      return false;
+    }
+
+    return true;
```
Comment (gregfrasco) [inline server/src/auth/user-permissions.service.ts:80]: Maybe the default should be false???

**Review by gregfrasco — APPROVED** [review]

File: server/src/app.module.ts (hunk 3 above)
Comment (cbolles) [reply server/src/app.module.ts:16]: Yes it should be, that'll be added in soon

File: server/src/auth/user-permissions.service.ts (hunk 5 above)
Comment (cbolles) [reply server/src/auth/user-permissions.service.ts:80]: Good point, it should be

File: middleware/sailauth.py (hunk 1 above)
Comment (cbolles) [reply middleware/sailauth.py:137]: On purpose for now just for rapid development. Once we find out how it'll be deployed we can add a config

File: middleware/sailauth.py (hunk 2 above)
Comment (cbolles) [reply middleware/sailauth.py:228]: Nope, I'll remove it

File: server/src/auth/dtos/resource.dto.ts (hunk 4 above)
Comment (cbolles) [reply server/src/auth/dtos/resource.dto.ts:4]: At first I was making a custom enum, then realized that it probably already existed lol

**Pull request merged**